lib/spack/llnl/util/filesystem.py

```diff
@@ -2080,6 +2096,154 @@ def find_libraries(libraries, root, shared=True, recursive=False):
 
 
+def find_all_shared_libraries(root, recursive=False):
+    """Convenience function that returns the list of all shared libraries found
+    in the directory passed as argument.
+
+    Args:
+        root (str): directory where to look recursively for library files
+        recursive (bool):
+    Returns:
+        List of all libraries found in ``root`` and subdirectories.
+    """
+    return find_libraries("*", root=root, shared=True, recursive=recursive)
+
+
+def find_all_static_libraries(root, recursive=False):
+    """Convenience function that returns the list of all static libraries found
+    in the directory passed as argument.
+
+    Args:
+        root (str): directory where to look recursively for library files
+        recursive (bool):
+    Returns:
+        List of all libraries found in ``root`` and subdirectories.
+    """
+    return find_libraries("*", root=root, shared=False, recursive=recursive)
+
+
+def find_all_libraries(root, recursive=False):
+    """Convenience function that returns the list of all libraries found
+    in the directory passed as argument.
+
+    Args:
+        root (str): directory where to look recursively for library files
+        recursive (bool):
+    Returns:
+        List of all libraries found in ``root`` and subdirectories.
+    """
+
+    return find_all_shared_libraries(root, recursive=recursive) + find_all_static_libraries(
+        root, recursive=recursive
+    )
+
+
+class WindowsSimulatedRuntimePath(object):
+    """Class representing Windows filesystem rpath analog
+
+    One instance of this class is associated with a package (only on Windows)
+    For each lib/binary directory in an associated package, this class introduces
+    a symlink to any/all dependent libraries/binaries. This includes the packages
+    own bin/lib directories, meaning the libraries are linked to the bianry directory
+    and vis versa.
+
+
+    """
+
+    def __init__(self, package, link_install_prefix=True):
+        """
+        Args:
+            package (spack.package_base.PackageBase): Package requiring links
+            link_install_prefix (bool): Link against package's own install or stage root.
+                Packages that run their own executables during build and require rpaths to
+                the build directory during build time require this option. Default: install
+                root
+        """
+        self.pkg = package
+        self._addl_rpaths = set()
+        self.link_install_prefix = link_install_prefix
+        self._internal_links = set()
+
+    @property
+    def link_dest(self):
+        """
+        Set of directories where package binaries/libraries are located.
+        """
+        if hasattr(self.pkg, "libs") and self.pkg.libs:
+            pkg_libs = set(self.pkg.libs.directories)
+        else:
+            pkg_libs = set((self.pkg.prefix.lib, self.pkg.prefix.lib64))
+
+        return pkg_libs | set([self.pkg.prefix.bin]) | self.internal_links
```

Comment:
In relation to (1) there's a couple ways I think this could be updated to avoid the need for `establish_internal_linkage`:

* Have this function look for libraries in `bin/` etc.
* Or we could update the implementation of `_libs_default_handler` to do this (and expect that custom overrides of `.libs` in packages would search `bin/` etc.)

If the files that need to symlink libraries are not strictly libraries themselves (i.e. other libraries/exe's don't need them) then the first option is probably better.

Reply:
Correct me if I misunderstood, but we do select the bin directory here for inclusion into the linking scheme.

Reply:
No, not at the moment. This is the scenario that I have been bringing up as a topic of concern. Trying to successfully balance digging into a package to find dependencies vs avoiding binaries/libraries that should not be linked against or overly verbose linking. I guess my thought is that this should follow similar rules as rpath/ld_library_path, i.e. not recursively dig into directories like `bin`.